hermes/src/network/rpc/ws.rs

```
@@ -0,0 +1,303 @@
+    let (tx, rx) = mpsc::channel::<Vec<PriceIdentifier>>(1000);
+
+    ws_state.subscribers.insert(id, tx);
+
+    log::debug!("New websocket connection, assigning id: {}", id);
+
+    let mut subscriber = Subscriber::new(id, state.store.clone(), rx, receiver, sender);
+
+    subscriber.run().await;
+}
+
+pub type SubscriberId = usize;
+
+/// Subscriber is an actor that handles a single websocket connection.
+/// It listens to the store for updates and sends them to the client.
+pub struct Subscriber {
+    id:                      SubscriberId,
+    closed:                  bool,
+    store:                   Store,
+    update_rx:               mpsc::Receiver<Vec<PriceIdentifier>>,
+    receiver:                SplitStream<WebSocket>,
+    sender:                  SplitSink<WebSocket, Message>,
+    price_feeds_with_config: HashMap<PriceIdentifier, PriceFeedClientConfig>,
+}
+
+impl Subscriber {
+    pub fn new(
+        id: SubscriberId,
+        store: Store,
+        update_rx: mpsc::Receiver<Vec<PriceIdentifier>>,
+        receiver: SplitStream<WebSocket>,
+        sender: SplitSink<WebSocket, Message>,
+    ) -> Self {
+        Self {
+            id,
+            closed: false,
+            store,
+            update_rx,
+            receiver,
+            sender,
+            price_feeds_with_config: HashMap::new(),
+        }
+    }
+
+    pub async fn run(&mut self) {
+        loop {
+            if self.closed {
+                break;
+            }
+            if let Err(e) = self.handle_next().await {
+                log::error!("Subscriber {}: Error handling next message: {}", self.id, e);
+                break;
+            }
+        }
+    }
+
+    async fn handle_next(&mut self) -> Result<()> {
+        tokio::select! {
+            Some(update_feed_ids) = self.update_rx.recv() => {
+                self.handle_price_feeds_update(update_feed_ids).await?;
+            },
+            Some(message_or_err) = self.receiver.next() => {
```

What circumstances does this return `None`? My understanding is that this will only occur when the stream ends but that likely means you want to return an error also?

---

I am handling it better now.